### DISCUSSION TITLE
Waypoint mission using Rosetta drone for dji mini 2

### DISCUSSION THREAD

**Discussion by andreamanera** [RosettaDrone/rosettadrone#163] [Q&A]
Hey everybody, I'm doing some research at my university in Italy, Genova, and we are interested in the world of UAV's, I'm new to this world and since we are interested in particular in waypoint navigation we came to this project, I'm learning now about how to create waypoint missions and in particular me and my team mates are requested to create an application to implement waypoint mission on DJI mini 2 drones, we discovered that this is not possible using the class WaypointMission from DJI so we are looking for some custom solution. Looking at this project we found out that you need to run QGC on your smartphone and set the waypoint on this application in order to have the waypoint mission working, now my question is do you think that it could be possible to implement the waypoint mission directly on rosetta drone without needing to run QGC? Thank you for your help!

**Comment by kripper** [comment]
There is a waypoint editor inside Rosetta, but I didn't add support for the MissionManager class (our custom waypoint implementation).
Adding support should be easy, but I won't do it myself, because I use QGC.

**Reply by andreamanera** [reply]
okay thanks for your help, and one more question could you please tell me, if you remember, at which script  I should look to add the support in the waypoint editor

**Comment by kripper** [comment]
No. Search the code for some string displayed in the user interface or the menu, and trace the code until you find the desired class.

If `MissionManager` is enabled (there is a variable for that in `DroneModel`), the internal waypoints editor should use the waypoints list from the `MissionManager`.

I remember there were two internal waypoints editors implementations and I just came and disabled the old implementation.
@The1only,  is there any feature or code we should port from the old version to the new version?
Why were there two implementations?

**Comment by The1only** [comment]
There was two ipmlementations as there was two examples in the DJI SDK and I hade problems desidig wich to use. Thow I belive only one actually was working. It was not used mush, more as a studdy to implement a better one, but I did no like the idea. For our usage an external laptop is way better.Best regardTerje Nilsen9Tek ASNorwayOn 18 May 2023, at 16:56, Christopher Pereira ***@***.***> wrote:﻿
No. Search the code for some string displayed in the user interface or the menu, and trace the code until you find the desired class.
If MissionManager is enabled (there is a variable for that in DroneModel), the internal waypoints editor should use the waypoints list from the MissionManager.
I remember there were two internal waypoints editors implementations and I just came and disabled the old implementation.
@The1only,  is there any feature or code we should port from the old version to the new version?
Why were there two implementations?

**Comment by kripper** [comment]
Agree. Having an internal waypoint editor is a bit reinventing the wheel when we can have QGC running on laptop and Android.
IMO, Rosetta should just provide the MAVLink wrapper.
Maybe QGC for Android can be started and better integrated with Rosetta for those users who are looking for an all in one app.

**Reply by andreamanera** [reply]
Yeah I get Your point, but we are basically aiming at what you mentioned, namely those users who are looking for an all in one app, and we noticed that there are no app who include all the functionalities, especially for those drones which didn't provide functionalities as the waypoint missions on board

**Comment by The1only** [comment]
You can use Android QGC on the same device, only it is a bit unpractikal. Switching between them can make strange things happening. Integrating them is quite a bit work, as QGC is written in C++/QML  and RosattaDrone in Java.Both are opensource thow.Best regardTerje Nilsen9Tek ASNorwayOn 18 May 2023, at 18:16, Andrea Manera ***@***.***> wrote:﻿
Yeah I get Your point, but we are basically aiming at what you mentioned, namely those users who are looking for an all in one app, and we noticed that there are no app who include all the functionalities, especially for those drones which didn't provide functionalities as the waypoint missions on board

**Comment by kripper** [comment]
Right. I would suggest evaluating if integrating QGC (opensource) with Rosetta could be a better option.

**Comment by kripper** [comment]
Sure. Buy if we describe what is currently "unpractical" and find some way to integrate QGC via command line arguments and find a solution for this use cases, we will have the best opensource waypoints editor (QGC) and have upstream support.

**Comment by kripper** [comment]
What do we need?

* Start QGC in the missing editor UI mode and download the current mission via MAVLINK (connect to local Rosetta MAVLINK connection)?

Otherwise, we would have to mimic all QGC features to have the best waypoints editor.

Rosetta would still provide all MAVLink and DJI SDK abstractions/wrappers. QGC can do the rest.

**Comment by The1only** [comment]
Then we are back to Rosettadrone 1 and the origional idea without GUI. It is a way easier path but quite different.Best regardTerje Nilsen9Tek ASNorwayOn 19 May 2023, at 01:02, Christopher Pereira ***@***.***> wrote:﻿
What do we need?

Start QGC in the missing editor UI mode and download the current mission via MAVLINK (connect to local Rosetta MAVLINK connection)?

Otherwise, we would have to mimic all QGC features to have the best waypoints editor.
Rosetta would still provide all MAVLink and DJI SDK abstractions/wrappers. QGC can do the rest.

**Comment by kripper** [comment]
I believe QGC for Android can be started from Rosetta like any other Rosetta activity (smoothly integrated) and be used as Rosetta's official waypoints editor. Thus Rosetta would still be rich in GUI and "all in one" app.

Any other effort to developed and maintain a self made waypoints editor in Java would be competing in disadvantage with QGC, which already has many features and is well maintained.

Besides, the two implementations we have in Rosetta are too coupled to the DJI SDK and don't really support all MAVLINK waypoint attributes.
For the MissionManager class, I used a list of MAVLink waypoints (supporting all MAVLink attributes).
I believe the editor should speak MAVLink too.

**Comment by The1only** [comment]
AgreeBest regardTerje Nilsen9Tek ASNorwayOn 19 May 2023, at 12:57, Christopher Pereira ***@***.***> wrote:﻿
I believe QGC for Android can be started from Rosetta like any other Rosetta activity (smoothly integrated) and be used as Rosetta's official waypoints editor. Thus Rosetta would still be rich in GUI and "all in one" app.
Any other effort to developed and maintain a self made waypoints editor in Java would be competing in disadvantage with QGC, which already has many features and is well maintained.
Besides, the two implementations we have in Rosetta are too coupled to the DJI SDK and don't really support all MAVLINK waypoint attributes.
For the MissionManager class, I used a list of MAVLink waypoints (supporting all MAVLink attributes).
I believe the editor should speak MAVLink too.
